fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/QueryTransformer.java

```diff
@@ -176,16 +175,7 @@ private OptExprBuilder filter(OptExprBuilder subOpt, Expr predicate) {
 
         SubqueryTransformer subqueryTransformer = new SubqueryTransformer(session);
         subOpt = subqueryTransformer.handleSubqueries(columnRefFactory, subOpt, predicate, cteContext);
-
-        ScalarOperator scalarPredicate =
-                subqueryTransformer.rewriteSubqueryScalarOperator(predicate, subOpt, correlation);
-
-        if (scalarPredicate != null) {
-            LogicalFilterOperator filterOperator = new LogicalFilterOperator(scalarPredicate);
-            return subOpt.withNewRoot(filterOperator);
-        } else {
-            return subOpt;
-        }
+        return subqueryTransformer.handlePredicate(subOpt, predicate, correlation);
```

Comment:
why only handle it in `filter-clause`? `projection` don't need? `join on`?

Reply:
Only where clause and having clause will come to this logic. On predicate is processed differently, it should be placed at the join operator not a separately filter operator

Reply:
Seems the logic shouldn't be placed here

Reply:
As SQL:
select if (1=2, (select t2 from xx), (select t1 from xx));

we can simplified the subquery, not only filter/having

Reply:
Well, I see. This is beyond my consideration, and I'll try it this way.